clang/lib/AST/Expr.cpp

```diff
@@ -1648,8 +1648,11 @@ SourceLocation CallExpr::getBeginLoc() const {
   if (const auto *Method =
           dyn_cast_if_present<const CXXMethodDecl>(getCalleeDecl());
       Method && Method->isExplicitObjectMemberFunction()) {
-    assert(getNumArgs() > 0 && getArg(0));
-    return getArg(0)->getBeginLoc();
+    if (!isTypeDependent()) {
+      assert(getNumArgs() > 0 && getArg(0));
+      if (getNumArgs() > 0 && getArg(0))
```

Comment:
In addition to fixing the reported test case by adding the `isTypeDependent()` check, I made this branch more defensive in general, by checking the condition `getNumArgs() > 0 && getArg(0)` after asserting it. That way, if there are additional edge cases that would fail this check, they still trip the assertion, but production builds avoid a crash in favour of a much more benign error (a suboptimal begin location).

Reply:
I don't think we do this elsewhere and in fact I think we prefer a crash we are violating an invariant and we want a hard fail so we can identify and fix ASAP.

Most folks don't run w/ assertions enabled and we could be hiding a latent bug for a long time.

Reply:
I agree with these arguments against checking an asserted condition **in general**.

However, I think in this case it might make sense, for two reasons:
 1. My level of confidence that "((callee is an explicit object member function) and (call is not dependent)) implies (call has a first argument)" is in fact an invariant is not very high.
 2. The impact of hiding a bug here is fairly low (slightly inaccurate source range, likely to only affect tools like clangd which perform hit-testing on the AST) compared to a production crash.

I'm happy to remove the check if you feel that's the right tradeoff in spite of these considerations, but I wanted to call them out.